#### Author: Rhys White ([@RhysTWhite](https://twitter.com/RhysTWhite))
#### Affiliations: Institute of Environmental Science and Research - ESR

# Welcome to the first Advanced Genomics Workshop<br> Hosted right here at ESR! 

Kia ora koutou! We are thrilled to guide you on your exciting journey into the realm of advanced genome bioinformatics. This hands-on course uses the command-line computing environment to help develop your skills in analysing genome sequence data.

## Course Description
The primary objective of this course is to equip you with the essential knowledge and practical computational skills required to perform diverse genome-scale analyses using data generated through next-generation sequencing (NGS) technologies. Through informative 'lectures', this course covers various topics, including read mapping, quality control, _de novo_ assembly, visualisation, and phylogenomics, providing comprehensive insights into the approaches for analysing genome data.

Emphasis is placed on fostering an understanding of the fundamental principles underlying commonly employed analysis tools.

You will engage in practical exercises throughout the course using authentic NGS data, Unix command-line tools, and high-performance computing resources. These computer-based activities will enable you to develop practical skills in genome informatics and gain proficiency in applying computational approaches to real-world genomic datasets.

## Before we start week 1
You will need a computer running a command line interface (e.g., macOS Terminal, Windows Command Prompt, Linux terminal). Participants in the course are expected to have a Linux account on the HPC (High-Performance Computing) system and have their secure shell (SSH) keys set up for server connection. 

**Please get in touch with Russell Smithies if you cannot access the HPC** - an email or message via Teams is sufficient. 

If you need help accessing the HPC, please refer to the Basic Server Usage FAQ on the ESR wiki page. Throughout the course, participants are expected to use their own work laptops. 

**If any software installation is required on laptops, it is advisable to coordinate with Desktop Support at least a week or two before the course. It is crucial to complete these steps beforehand. Performing these tasks on the morning of the course is not recommended.**

Before delving into genomic sequence data analysis, week 1 will ensure everyone's comfort with Unix commands. Throughout the first two weeks, there will be opportunities to discuss and address any challenges encountered with specific commands. However, should you encounter difficulties in week 1, we recommend thoroughly exploring the Unix tutorial notes in the Software Carpentry resources. You can access them here: [https://swcarpentry.github.io/shell-novice/index.html](https://swcarpentry.github.io/shell-novice/index.html)

## What we will cover?

The topic covered are as follows:

Week 1: An introduction to the command line<br>
Week 2: Conquering the command line with confidence<br>
Week 3: Navigating read-mapping files<br>
Week 4: Short-read sequence data quality control<br>
Week 5: Long-read sequence data quality control<br>
Week 6: Assembling a genome using sequence data<br>
_Week 7: Multi-locus sequencing typing, antibiotic resistance gene detection<br>
Week 8: Phylogenetics<br>
Week 9: Amplicon sequencing<br>
Week 10: Using Jupyter Notebooks_


## A final note...
Everyone works at their own pace, so it's important to recognise that setting aside an extra couple of hours per week (aside from the hands-on tutorials) might be beneficial to maximise your progress.

We hope you enjoy this learning experience and find it rewarding.

NgƒÅ mihi,

Rhys White <sub> PhD </sub> (Course Coordinator)<br>
ESR Scientist


On behalf of the Bioinformatics and Human Genomics team

![image](https://github.com/ESR-NZ/Training/assets/28285670/52788e9c-0316-44f0-93ca-4e27ea3fd323)

