ONT sequence data QC
#### Author: Rhys White ([@RhysTWhite](https://twitter.com/RhysTWhite))
#### Affiliations: Institute of Environmental Science and Research - ESR

# Advanced Genomic Insights Workshop Week 05:<br> A tutorial on long-read sequence data quality control

Table of contents
=================

<!--ts-->
   * [Prerequisites](#prerequisites)
   * [Introduction](#introduction)
   * [Let's get started...](#lets-get-started)
      * [Analysing Quality of Sequence Read Data with FastQC](#analysing-quality-of-sequence-read-data-with-fastqc)
      * [Sequence Data Processing with Trimmomatic: Quality Filtering and Adapter Removal](#sequence-data-processing-with-trimmomatic-quality-filtering-and-adapter-removal)
      * [Conquering Taxonomic Classification with Kraken2](#conquering-taxonomic-classification-with-kraken2)
      
      
   * [Why have I learnt this?](#why-have-i-learnt-this)
   * [Final notes](#final-notes)
   * [Data availability](#data-availability)
<!--te-->

## Prerequisites
To fully benefit from these exercises, it is essential that you manually _type_ each exercise. Please avoid the temptation to copy and paste, as these exercises aim to train your hands, brain, and mind in the fundamental skills of reading, writing, and comprehending code. By copying and pasting, you would be depriving yourself of the valuable learning experience these lessons provide.

Before we begin, you should have a basic understanding of genomics and the concept of DNA sequencing. You should also be familiar with the command line and have access to a Unix-based operating system (such as Linux or macOS).

## Introduction

In the world of genomics and bioinformatics, it is crucial to ensure the quality of the sequencing data before proceeding with any downstream analyses. In this tutorial, you will receive a step-by-step guide on utilising three 'gold-standard' tools to guarantee that your Oxford Nanopore Technologies (ONT) sequencing data is of high quality before beginning your analyses.

## Let's get started...

First, ensure that you are currently located within the `week05` directory, which you can create as per the instructions provided in [Week 01: an introduction to the command line](https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_01).

##### Getting the data
To begin, create a directory called `raw_reads`. Afterwards, navigate to the `raw_reads` folder:

```bash
# Make a directory called raw_reads:
mkdir raw_reads
```

```bash
# Change directory to your raw_reads directory
cd  raw_reads
```

Note: The data is ~236 MB. To download the data, please make a script file named `wget_download.sh` using a text editor:

```bash
# Download the files with ONT raw reads:
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_0.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_1.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_2.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_3.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_4.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_5.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_6.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_7.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_8.fastq.gz
wget https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05/Fastq_pass_barcode11/FAS91936_pass_barcode11_069450e3_1619d8e7_9.fastq.gz
```

Now, make sure you are in the `raw_reads` folder (as well as your `wget_download.sh` script). Please run the download using:

```bash
# Run your wget_download.sh script:
bash wget_download.sh
```

You have been given ten `fastq.gz` files. Each file is a compressed text file containing DNA sequence data and corresponding quality scores generated by ONT sequencing machines.

To confirm that all ten files are present in your current working directory, check for their existence using the following command:

```bash
# Directory listing
ls
```

### Analysing Quality of Sequence Read Data with nanoQC
