#### Author: Rhys White ([@RhysTWhite](https://twitter.com/RhysTWhite))
#### Affiliations: Institute of Environmental Science and Research - ESR

# Advanced Genomic Insights Workshop Week 07:<br> A tutorial on assembling a genome using long-read sequence data

Table of contents
=================

<!--ts-->
   * [Prerequisites](#prerequisites)
   * [Introduction](#introduction)
   * [Let's get started...](#lets-get-started)
      * [Read _De Novo_ Assembly](#read-de-novo-assembly)
      * [_De novo_ assembly of nanopore sequence reads with Flye](#de-novo-assembly-of-nanopore-sequence-reads-with-flye)
      * [Analyse the assembly metrics](#analyse-the-assembly-metrics)      
      * [Conquering Taxonomic Classification with Kraken2](#conquering-taxonomic-classification-with-kraken2)  
   * [Why have I learnt this?](#why-have-i-learnt-this)
   * [Final notes](#final-notes)
   * [Data availability](#data-availability)
<!--te-->

## Prerequisites
To fully benefit from these exercises, it is essential that you manually _type_ each exercise. Please avoid the temptation to copy and paste, as these exercises aim to train your hands, brain, and mind in the fundamental skills of reading, writing, and comprehending code. By copying and pasting, you would be depriving yourself of the valuable learning experience these lessons provide.

Before we begin, you should have a basic understanding of genomics and the concept of DNA sequencing. You should also be familiar with the command line and have access to a Unix-based operating system (such as Linux or macOS).

Prior completion of the week 05 sequence data [quality control tutorial](https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05) is needed. We will analyse the same sequence data in that tutorial for time efficiency.

## Introduction

_De novo_ assembly of nanopore sequence read data involves piecing long DNA sequences (reads) into longer contiguous sequences (contigs) without a reference genome (i.e., without any prior knowledge of their correct sequence or order). Nanopore sequencing offers several advantages, such as long read lengths and real-time sequencing. Still, it also has limitations and specific issues that must be addressed during the assembly process. This tutorial will review the basic _de novo_ assembly of nanopore sequence read data.

## Let's get started...

##### Getting the data

Note: For the [quality control tutorial](https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05), you would have downloaded 10 `fastq.gz` files, and then concatenate them into a single `.fastq.gz` file (the suggested name for this concatenated file was `22AR0430_raw_ONT.fastq.gz`. To recap, this `22AR0430_raw_ONT.fastq.gz` file is a compressed text file containing DNA sequence data and corresponding quality scores generated by Nanopore sequencing machines.

You would have investigated the quality of your sequence read data using `nanoQC` and successfully used `NanoFilt` to remove low-quality bases and Illumina adaptor sequences from your reads. 

**The trimmed/filtered reads are what we will be using today.**

##### Creating symbolic links: A step-by-step guide on the command line

Creating soft links, or symbolic links, is a way to reference a file or directory in another location on a Linux or Unix system. Soft links are helpful when creating a shortcut to access a file or directory without making a duplicate copy. This saves space on our system and provides a more efficient method of managing files and directories.

To begin, create a directory called `assembly`. Afterwards, navigate to the `assembly` folder:

```bash
# Make a directory called assembly:
mkdir ONT_assembly
```

```bash
# Change directory to your assembly directory
cd ONT_assembly
```

###### Create the soft link

To create a soft link, use the `ln` command with the `-s` option, followed by the path to the file or directory you want to link, and the name of the soft link you wish to create. Edit this command to create a soft link to the trimmed reads from the [quality control tutorial](https://github.com/ESR-NZ/Training/tree/main/Advanced_Genomics_Workshop/Week_05):

```bash
# Create a soft link to the file with nanopore single-ended reads:
ln -s /path/to/filtered_reads/${SAMPLE}_filtered_ONT.fastq.gz .
```

Replace `${SAMPLE}_filtered_ONT.fastq.gz` with the name of your input file.

To confirm that the file is present in your current working directory, check for their existence using the following command:

```bash
# Directory listing
ls
```

To obtain more information about the `ln` command and its available options, you can execute the command `ln --help`.

### Read _De Novo_ Assembly

The next step is assembling the nanopore reads into contigs using a _de novo_ assembler. Many _de novo_ assemblers are available, but today we will focus on Flye ([Lin et al. 2016](
https://doi.org/10.1073/pnas.1604560113); [Kolmogorov et al. 2020](https://doi.org/10.1038/s41587-019-0072-8)).

Remember to cite the paper:

&emsp; Kolmogorov M, Yuan J, Lin Y, Pevzner PA. Assembly of long, error-prone reads using repeat graphs. _Nature Biotechnology_ 2019;37:540-546 doi [10.1038/s41587-019-0072-8](https://doi.org/10.1038/s41587-019-0072-8)

### _De novo_ assembly of nanopore sequence reads with Flye





ONT sequence data Assembly


Notes
medaka_consensus -d .fasta -i ONT.fastq.gz -o medaka_test
